spec/src/main/asciidoc/jsonb.adoc

```diff
@@ -318,7 +318,54 @@ Deserialization into anonymous classes is not supported. Serialization of anonym
 
 === Polymorphic Types
 
-Deserialization into polymorphic types is not supported by default mapping.
+Polymorphic type handling is supported for deserialization and serialization. Polymorphic handling is ensured by annotation `@JsonbPolymorphicType` and `@JsonbSubtype`. `@JsonbPolymorphicType` defines overall strategy how polymorphic information will be serialized/deserialized and defines all the supported aliases with using `@JsonbSubtype` annotations. `@JsonbSubtype` ensures proper and safe mapping between class alias and type. Implementation must validate mapped types if they are assignable from the annotated type. If not, an exception must be thrown.
+
+Polymorphic information is obtained from `@JsonbSubtype` annotation as a type alias mapped to the type or if `classNames` property of the `@JsonbPolymorphicType` is enabled, it corresponds to the fully qualified class name, if no suitable alias is found. Implementation must ensure to support `allowedPackages` option of the `@JsonbPolymorphicType` which is only used when `classNames` option is enabled, it is ignored otherwise. It is not allowed to create instances of the classes outside the allowed packages. In such cases an exception must be thrown. If allowed packages are empty, all classes are allowed.
+
+Available serialization formats:
+
+* Property - new property with polymorphic information is added to the serialized object. The property key name is taken from the `key` property of the annotation `@JsonbPolymorphicType`. This polymorphic information property key name has to be unique in the resulting JSON document. If any naming collision with class properties occurs, an exception must be thrown. It is required for all polymorphism fields to be serialized as the first properties in the json and an actual object properties are serialized after.
+* Wrapping object - serialized object is wrapped with another object with a single property. The key corresponds to the polymorphic information of the type and the value is the serialized object
+
+If no `@JsonbPolymorphicType` is used on handled class or its predecessors, it is not possible to ensure proper polymorphic handling and in such cases deserialization is not supported for polymorphic types.
+
+If the type predecessor has set `@JsonbPolymorphicType` and the current type also, but with the different format, an exception must be thrown.
+
+If there are multiple of different polymorphic customizations, which need to be merged (such as class has predecessor with `@JsonbPolymorphicType` and one of its interfaces does have `@JsonbPolymorphicType` set also, or two or more interfaces do have `@JsonbPolymorphicType` set) an exception must be thrown. See example below.
```

Comment:
If there are multiple different polymorphic customizations that need to be merged, an exception must be thrown. For example, if class has a predecessor with `@JsonbPolymorphicType` and one of its interfaces also has `@JsonbPolymorphicType` set, or two or more interfaces do have `@JsonbPolymorphicType` set, then an exception must be thrown.